neuralpredictors/layers/cores/conv2d.py

```diff
@@ -292,122 +310,97 @@ def __init__(
-            BiasLayer = Bias2DLayer
-            ScaleLayer = Scale2DLayer
+    def bias_layer(self, **kwargs):
+        return RotationEquivariantBias2DLayer(num_rotations=self.num_rotations, **kwargs)
 
-        if use_avg_reg:
-            warnings.warn("The averaged value of regularizer will be used.", UserWarning)
-
-        self.features = nn.Sequential()
-        if stack is None:
-            self.stack = range(self.layers)
-        else:
-            self.stack = [*range(self.layers)[stack:]] if isinstance(stack, int) else stack
+    def scale_layer(self, **kwargs):
+        return RotationEquivariantScale2DLayer(num_rotations=self.num_rotations, **kwargs)
 
-        # --- first layer
+    def add_first_layer(self):
         layer = OrderedDict()
         layer["conv"] = HermiteConv2D(
-            input_features=input_channels,
-            output_features=hidden_channels,
-            num_rotations=num_rotations,
-            upsampling=upsampling,
-            filter_size=input_kern,
-            stride=stride,
-            padding=input_kern // 2 if pad_input else 0,
+            input_features=self.input_channels,
+            output_features=self.hidden_channels,
+            num_rotations=self.num_rotations,
+            upsampling=self.upsampling,
+            filter_size=self.input_kern,
+            stride=self.stride,
+            padding=self.input_kern // 2 if self.pad_input else 0,
             first_layer=True,
         )
-        if batch_norm:
-            if independent_bn_bias:
-                layer["norm"] = BatchNormLayer(num_features=hidden_channels, momentum=momentum)
+        if self.batch_norm:
+            if self.independent_bn_bias:
+                layer["norm"] = self.batchnormlayer_cls(num_features=self.hidden_channels, momentum=self.momentum)
             else:
-                layer["norm"] = BatchNormLayer(
-                    num_features=hidden_channels,
-                    momentum=momentum,
-                    affine=bias and batch_norm_scale,
+                layer["norm"] = self.batchnormlayer_cls(
+                    num_features=self.hidden_channels,
+                    momentum=self.momentum,
+                    affine=self.bias and self.batch_norm_scale,
                 )
-                if bias:
-                    if not batch_norm_scale:
-                        layer["bias"] = BiasLayer(channels=hidden_channels)
-                elif batch_norm_scale:
-                    layer["scale"] = ScaleLayer(channels=hidden_channels)
-
-        if layers > 1 or final_nonlinearity:
-            layer["nonlin"] = AdaptiveELU(elu_xshift, elu_yshift)
+                if self.bias:
+                    if not self.batch_norm_scale:
+                        layer["bias"] = self.biaslayer_cls(channels=self.hidden_channels)
+                elif self.batch_norm_scale:
+                    layer["scale"] = self.scalelayer_cls(channels=self.hidden_channels)
```

Comment:
This block of code is also shared similarly to the one in `Stacked2dCore`.

Reply:
yes very good point. I've refactored it. Do you think it makes sense now?

Reply:
Yes much better but I think we can do even better. I think you can use the new `add_activation` method in `RotationEquivariant2dCore` as well. 
Also the code that adds the batch norm layer in `RotationEquivariant2dCore` is almost the same as the one in `Stacked2dCore`. The only difference is that `RotationEquivariant2dCore` uses instance variables for the different layer classes instead of using the classes directly like `Stacked2dCore` does. Could we also use variables in the `add_bn_layer` method? Then we could use that method in `RotationEquivariant2dCore` as well.